src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
+              if(originalID == FieldType::FT_NotOriginal)
+                continue;
+              auto idWithExtents = dirtyFields.find(originalID);
+              if(idWithExtents == dirtyFields.end())
+                continue;
+              // If the access is horizontally pointwise or it is a horizontally cached field,
+              // we do not need to trigger a BC
+              if(readacccess.second.isHorizontalPointwise() ||
+                 stencilInstantiation->getCachedVariableSet().count(readacccess.first))
+                continue;
+              auto finder = allBCs.find(originalID);
+              // Check if a boundary condition for this variable was defined
+              if(finder == allBCs.end()) {
+                DAWN_ASSERT_MSG(
+                    false,
+                    dawn::format("In stencil %s we need a halo update on field %s but no "
+                                 "boundary condition is set.\nUpdate the stencil (outside the "
+                                 "do-method) with a boundary condition that calls a "
+                                 "stencil_function, e.g \n'boundary_condition(zero(), %s);'\n",
+                                 stencilInstantiation->getName(),
+                                 stencilInstantiation->getOriginalNameFromAccessID(originalID),
+                                 stencilInstantiation->getOriginalNameFromAccessID(originalID))
+                        .c_str());
+              }
+              // Calculate the extent and add it to the boundary-condition - Extent map
+              Extents fullExtents = calculateHaloExtents(
+                  stencilInstantiation->getNameFromAccessID(readacccess.first));
+              stencilInstantiation->getBoundaryConditionToExtentsMap().emplace(finder->second,
+                                                                               fullExtents);
+
+              // check if this stencil is called and get its StencilCallDeclStmt (the one to replace)
+              auto test =
+                  stencilInstantiation->getIDToStencilCallMap().find(stencil.getStencilID());
+              if(test == stencilInstantiation->getIDToStencilCallMap().end()) {
+                DAWN_ASSERT_MSG(false, "Stencil Triggering the Boundary Condition is not called");
```

Comment:
use count(key)
I dont understand this assertion, can you explain it? Should there be a map 1 to 1 between stencil call and the stencils of stencil instantiation?